### PR TITLE
Rename `Struct` -> `LegacyStruct` and misc cleanup

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Rename `Struct` -> `LegacyStruct`, this class template will be deprecated soon. Also, created a `test/test_polymorphic_value.cpp` for polymorphic value tests. Currently, there is only one test `OpaqueEquality` moved from `test/test_evaluator.cpp`, in the future, more will be added.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -237,7 +237,7 @@ if(PROJECT_IS_TOP_LEVEL)
 
   # this is to find pip installed nvrtc/nvtx .so
   set_target_properties(${NVFUSER_CODEGEN} PROPERTIES INSTALL_RPATH
-	  "$ORIGIN/../../nvidia/cuda_nvrtc/lib:$ORIGIN/../../nvidia/nvtx/lib")
+    "$ORIGIN/../../nvidia/cuda_nvrtc/lib:$ORIGIN/../../nvidia/nvtx/lib")
```
Comment (zasdfgbnm) [inline CMakeLists.txt:240]: Formatted by auto formatter

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Review by csarofeen — APPROVED** [review]
LGTM

**Pull request merged**